NetJSON/NetJSON.cs

```diff
@@ -3727,80 +3727,87 @@ public static object Deserialize(Type type, string value, NetJSONSettings settin
-                                    var uu = Int32.Parse(str, NumberStyles.HexNumber);
-                                    
-                                    if(uu < offset)
-                                    {
-                                        sb.Append((char)uu);
-                                    }
-                                    else
-                                    {
-                                        sb.Append((char)(((uu - offset) >> 10) + 0xD800))
-                                            .Append((char)((uu - offset) % 0x0400 + 0xDC00));
-                                    }
-
-                                    index += 4;
-                                    break;
-                                default:
-                                    if (next == settings._quoteChar)
-                                        sb.Append(next);
-                                    break;
-                            }
-                        }
-                    }
-                } else {
+						break;
+					}
+
+					if (current != '\\') {
+						sb.Append(current);
+						prev = current;
+						++ptr;
+						continue;
+					}
+
+					next = *(++ptr);
+					switch (next) {
+						case '\0': goto EXIT; // string ends with '\'
+						case 'r': sb.Append('\r'); break;
+						case 'n': sb.Append('\n'); break;
+						case 't': sb.Append('\t'); break;
+						case 'f': sb.Append('\f'); break;
+						case '\\': sb.Append('\\'); break;
+						case '/': sb.Append('/'); break;
+						case 'b': sb.Append('\b'); break;
+						case 'u':
+							const int offset = 0x10000;
+							int uu = 0;
+							for (int i = 0; i < 4; i++) {
+								;
+								var c = *(++ptr);
+								if (c >= '0' && c <= '9') {
+									uu = (uu <<= 4) + (c - '0');
+								}
+								else if (c >= 'a' && c <= 'f') {
+									uu = (uu <<= 4) + c - ('a' - 10);
+								}
+								else if (c >= 'A' && c <= 'F') {
+									uu = (uu <<= 4) + c - ('A' - 10);
+								}
+								else {
+									throw new ArgumentException("Invalid Unicode escape sequence");
+								}
+							}
+
+							if (uu < offset) {
+								sb.Append((char)uu);
+							}
+							else {
+								sb.Append((char)(((uu - offset) >> 10) + 0xD800))
+									.Append((char)((uu - offset) % 0x0400 + 0xDC00));
+							}
+							break;
+						default:
+							if (next == settings._quoteChar)
+								sb.Append(next);
+							break;
+					}
+				} else {
                     if (current == settings._quoteChar) {
                         hasQuote = true;
                     } else if (current == 'n') {
-                        index += 3;
-                        return null;
+						ptr += 3;
+						index += (int)(ptr - startPtr);
+						return null;
                     }
                 }
                 prev = current;
-                ++index;
+				++ptr;
             }
-
+			EXIT:
+			index += (int)(ptr - startPtr);
```

Comment:
This is fine since it will add the changes between the original pointer and what it currently points to.